kratos/includes/global_pointer.h

```diff
@@ -0,0 +1,186 @@
+    , mRank(GetLocalRank())
+#endif
+    {
+  }
+
+  /** Constructor by boost::shared_ptr
+   * Constructor by boost::shared_ptr
+   * @param DataPointer Boost Shared Pointer to the Data.
+   */
+  GlobalPointer(boost::shared_ptr<TDataType> DataPointer)
+    : mDataPointer(DataPointer.get())
+#ifdef KRATOS_USING_MPI
+    , mRank(GetLocalRank())
+#endif
+    {
+  }
+
+  /** Constructor by boost::weak_ptr
+   * Constructor by boost::weak_ptr
+   * @param DataPointer Boost Weak Pointer to the Data.
+   */
+  GlobalPointer(boost::weak_ptr<TDataType> DataPointer) = delete;
+
+  /** Constructor by boost::unique_ptr
+   * Constructor by boost::unique_ptr
+   * @param DataPointer Std Unique Pointer to the Data.
+   */
+  /// Note: Not currently even imported on kratos.
+  // GlobalPointer(boost::unique_ptr<TDataType> DataPointer) = delete;
+
+  /** Constructor by std::shared_ptr
+   * Constructor by std::shared_ptr
+   * @param DataPointer Std Shared Pointer to the Data.
+   */
+  GlobalPointer(std::shared_ptr<TDataType> DataPointer) = delete;
+
+  /** Constructor by std::weak_ptr
+   * Constructor by std::weak_ptr
+   * @param DataPointer Std Weak Pointer to the Data.
+   */
+  GlobalPointer(std::weak_ptr<TDataType> DataPointer) = delete;
+
+  /** Constructor by std::unique_ptr
+   * Constructor by std::unique_ptr
+   * @param DataPointer Std Unique Pointer to the Data.
+   */
+  GlobalPointer(std::unique_ptr<TDataType> DataPointer) = delete;
+
+
+  /** Default Destructor
+   * Default Destructor.
+   */
+  ~GlobalPointer() {
+  }
+
+  /**
+  * Pointer Operator
+  */
+  TDataType & operator*() {
+	  return *mDataPointer;
+  }
+
+  /**
+  * Pointer Operator
+  */
+  TDataType const& operator*() const {
+	  return *mDataPointer;
+  }
+
+  /**
+   * Arrow Operator
+   */
+  TDataType * operator->() {
+    return mDataPointer;
+  }
+
+  /** Returns the rank of the data owner
+   * Returns the rank of the data owner
+   * @return rank of the data owner
+   */
+  int GetRank() const {
+#ifdef KRATOS_USING_MPI
+    return mRank;
+#else
+    return 0;
+#endif
+  }
+
+  /** Fills buffer with the GlobalPoiter data
+   * Fills buffer with the GlobalPoiter data
+   * @param buffer Object data buffer
+   */
+  void Save(char * buffer) {
+    memcpy(buffer, this, sizeof(GlobalPointer));
+  }
+
+  /** Restores the GlobalPoiter with the data from the buffer
+   * Restores the GlobalPoiter with the data from the buffer
+   * @param buffer Object data buffer
+   */
+  void Load(char * buffer) {
+    memcpy(this, buffer, sizeof(GlobalPointer));
+  }
+
+private:
+  /** Returns the rank of the current process
+   * Returns the rank of the current process
+   * @return Rank of the current process
+   */
+  int GetLocalRank() const {
+#ifdef KRATOS_USING_MPI
+    int mpi_rank;
+
+    MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
+
+    return mpi_rank;
+#else
+    return 0;
+#endif
+  }
+
+};
+
+} // namespace Kratos
+
+#endif // KRATOS_GLOBAL_POINTER_H_INCLUDED
```

Comment:
Would you please add the copy constructor, move constructor and copy and move = operator?